tests/handlers/test_directory.py

```diff
@@ -432,6 +432,60 @@ def test_allowed(self):
         self.assertEquals(200, channel.code, channel.result)
 
 
+class TestCreatePublishedRoomACL(unittest.HomeserverTestCase):
+    user_id = "@admin:test"
+    denied_user_id = "@test:test"
+    data = {"room_alias_name": "unofficial_test"}
+
+    servlets = [directory.register_servlets, room.register_servlets]
+
+    def prepare(self, reactor, clock, hs):
+        # This time we add custom room list publication rules
+        config = {}
+        config["alias_creation_rules"] = []
+        config["room_list_publication_rules"] = [
+            {"user_id": "*", "alias": "*", "action": "deny"},
+            {"user_id": "@admin:test", "alias": "*", "action": "allow"},
+        ]
+
+        rd_config = RoomDirectoryConfig()
+        rd_config.read_config(config)
+
+        self.hs.config.roomdirectory.is_publishing_room_allowed = (
+            rd_config.is_publishing_room_allowed
+        )
+
+        return hs
+
+    def test_denied(self):
+        # NOTE Setting is_public=True isn't enough
+        self.data["visibility"] = "public"
+        self.helper.create_room_as(
+            self.denied_user_id, extra_content=self.data, expect_code=403
+        )
```

Comment:
This is going to mutate `self.data`, which might affect tests called afterwards. That means that tests can be sensitive to the order they're ran in, or might not run the same in isolation!

For this specific case, I think the other calls to `create_room_as` explicitly set `is_public=False` and so each test will behave the same no matter which order they're ran in. But I think we should try to avoid mutating `self.data` anyway, so the reader doesn't have to reason through this.

How about:

```suggestion
        data = self.data.copy()
        data["visibility"] = "public"
        self.helper.create_room_as(
            self.denied_user_id, extra_content=data, expect_code=403
        )
```

Reply:
> This is going to mutate self.data, which might affect tests called afterwards. 

I'm not sure this is correct: I think there will be a new instance of `TestCreatePublishedRoomACL` for each test.

Reply:
ohhh, but each instance will end up pointing at the same `dict`, I guess, because it's a class-level property rather than an instance-level one. Maybe it'd be better to intialise `self.data` in `prepare`?

Reply:
> I'm not sure this is correct: I think there will be a new instance of TestCreatePublishedRoomACL for each test.

Looks like it! See https://gist.github.com/DMRobertson/40fe186843a327658cbee31bb381e1b2